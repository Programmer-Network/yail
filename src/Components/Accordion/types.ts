import { IDraggableListItem } from "../DraggableList/types";

export interface IAccordionSectionItem {
  id: number;
  title: string;
  order: number;
}

export interface ISection {
  id: number;
  description?: string;
  title: string;
  order: number;
  items: IAccordionSectionItem[];
}

export enum AccordionOrderType {
  SECTIONS = "sections",
  LECTURES = "lectures"
}

export interface IAccordionProps {
  className?: string;
  sections: ISection[];
  sectionTitleClassName?: string;
  onSorted?: (
    type: AccordionOrderType,
    sections: ISection[],
    section: { sectionId: number; items: IDraggableListItem[] }
  ) => void;
  onSectionItemClick?: (item: IDraggableListItem) => void;
  onSectionClick?: (item: IDraggableListItem) => void;
  onSelected?: (item: IDraggableListItem) => void;
  onAddSection?: () => void;
  addSectionItemLabel?: string;
  onAddSectionItem?: (item: ISection) => void;
  addSectionLabel?: string;
  expanded: number[];
  setExpanded: (sections: number[]) => void;
  hasDraggableSections?: boolean;
  hasDraggableSectionItems?: boolean;
  defaultExpanded?: boolean;
  selected?: {
    sectionId: number | null;
    itemId: number | null;
  };
  setSelectedSection?: (sectionId: number | null) => void;
  setSelectedItem?: (itemId: number | null) => void;
}
