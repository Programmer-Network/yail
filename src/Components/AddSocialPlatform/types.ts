export interface IAddSocialPlatformProps {
  className?: string;
  platforms: { name: string; url: string }[];
  onAdd: (data: Record<string, string>) => void;
}
