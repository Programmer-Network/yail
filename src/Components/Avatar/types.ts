export interface IAvatarProps {
  src?: string;
  alt?: string;
  size?: number;
  colors?: string[];
  className?: string;
}
