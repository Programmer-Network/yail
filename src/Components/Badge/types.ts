export enum BadgeVariantEnum {
  PRIMARY = "primary",
  SECONDARY = "secondary",
  SUCCESS = "success",
  WARNING = "warning",
  ERROR = "error",
  BETA = "beta",
  NEW = "new"
}

export interface IBadgeProps {
  title: string;
  className?: string;
  variant: BadgeVariantEnum;
}
