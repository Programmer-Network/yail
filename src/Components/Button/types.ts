import { ReactNode } from "react";

import { IconName } from "Components/Icons/types";

export enum ButtonVariantEnum {
  PRIMARY = "primary",
  SECONDARY = "secondary"
}

export interface IButtonProps {
  onClick?: (e: React.MouseEvent) => void;
  children?: ReactNode;
  className?: string;
  disabled?: boolean;
  type?: "button" | "submit" | "reset";
  isLoading?: boolean;
  outlined?: boolean;
  icon?: {
    iconName?: IconName;
    iconClassName?: string;
    iconPosition?: "left" | "right";
  };
  variant?: ButtonVariantEnum;
}
