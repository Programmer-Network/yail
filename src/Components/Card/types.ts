import { NavLinkProps } from "react-router-dom";

import { PillSize, PillVariant } from "../Pill/Pill.types";

export interface ICardTag {
  name: string;
  url: string;
}

export interface IAuthorProps {
  name: string;
  url?: string;
  avatarUrl?: string;
}

export interface IImageProps {
  src: string;
  alt: string;
  aspectRatio?: "16:9" | "4:3" | "1:1" | "3:2" | "auto";
  lazy?: boolean;
}

export interface IPill {
  title: string;
  variant?: PillVariant;
  className?: string;
  size?: PillSize;
  clickable?: boolean;
  onClick?: (e: React.MouseEvent) => void;
}

export interface ICardData {
  title: string;
  titleUrl: string;
  description: string;
  date?: string;
  author?: IAuthorProps;
  tags?: ICardTag[];
  pills?: IPill[];
  image?: IImageProps;
  isRead?: boolean;
  isFeatured?: boolean;
}

export interface ICardAction {
  label: string;
  onClick: (e: React.MouseEvent) => void;
  variant?: "primary" | "danger" | "outlined";
  icon?: React.ReactNode;
  show?: boolean; // for conditional rendering
}

export interface ICard {
  data: ICardData;
  NavLink: React.ForwardRefExoticComponent<
    NavLinkProps & React.RefAttributes<HTMLAnchorElement>
  >;
  className?: string;
  variant?: "default" | "compact" | "featured";
  isLoading?: boolean;
  maxVisibleTags?: number;
  maxTitleLines?: number;
  maxDescriptionLines?: number;
  showBookmark?: boolean;
  isBookmarked?: boolean;
  showShare?: boolean;
  onCardClick?: (data: ICardData) => void;
  onAuthorClick?: (author: IAuthorProps) => void;
  onTagClick?: (tag: ICardTag) => void;
  onImageClick?: () => void;
  onBookmark?: (data: ICardData) => void;
  onShare?: (data: ICardData) => void;
  actions?: ICardAction[];
  footerActions?: ICardAction[];
}

export interface ICardSkeletonProps {
  className?: string;
  variant?: "default" | "compact" | "featured";
  showImage?: boolean;
}
