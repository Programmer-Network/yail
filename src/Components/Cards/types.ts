import { NavLinkProps } from "react-router-dom";

import { ICardData } from "../Card/types";

export interface ICardsProps {
  cards: ICardData[];
  NavLink: React.ForwardRefExoticComponent<
    NavLinkProps & React.RefAttributes<HTMLAnchorElement>
  >;
  columns: number;
  className?: string;
  cardClassName?: string;
}
