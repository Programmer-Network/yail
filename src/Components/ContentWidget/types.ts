export interface IContentWidgetBaseItem {
  id?: number;
  title: string;
}

export interface IContentWidgetItemMore {
  tooltipText: string;
  iconName?: string;
  iconClassName?: string;
  onClick: () => void;
}

export interface IContentWidgetItem<T extends IContentWidgetBaseItem> {
  item: T;
  onClick: (item: T) => void;
}

export interface IContentWidgetProps<T extends IContentWidgetBaseItem> {
  title: string;
  className?: string;
  isLoading: boolean;
  items: IContentWidgetItem<T>[];
  more?: IContentWidgetItemMore;
}
