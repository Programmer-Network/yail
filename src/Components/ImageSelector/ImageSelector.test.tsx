import { fireEvent, render, screen } from "@testing-library/react";
import { vi } from "vitest";

import { ImageSelector } from "./";
import { IImage } from "./types";

describe("ImageSelector component", () => {
  const mockImages: IImage[] = [
    { id: 1, url: "http://example.com/image1.jpg" },
    { id: 2, url: "http://example.com/image2.jpg" }
  ];

  test("renders correctly - snapshot test", () => {
    const { asFragment } = render(
      <ImageSelector
        images={mockImages}
        value={mockImages[0]!}
        onSelected={vi.fn()}
        onDelete={vi.fn()}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("renders images correctly", () => {
    render(
      <ImageSelector
        images={mockImages}
        value={mockImages[0]!}
        onSelected={vi.fn()}
        onDelete={vi.fn()}
      />
    );
    mockImages.forEach(image => {
      expect(screen.getByTestId(`image-${image.id}`)).toBeInTheDocument();
    });
  });

  test("handles image selection", () => {
    const onSelected = vi.fn();
    render(
      <ImageSelector
        images={mockImages}
        value={mockImages[0]!}
        onSelected={onSelected}
        onDelete={vi.fn()}
      />
    );
    const secondImage = mockImages[1]!;
    fireEvent.click(screen.getByTestId(`image-${secondImage.id}`));
    expect(onSelected).toHaveBeenCalledWith(secondImage);
  });

  test("handles image deletion", async () => {
    const onDelete = vi.fn().mockResolvedValue([]);
    const firstImage = mockImages[0]!;
    render(
      <ImageSelector
        images={mockImages}
        value={firstImage}
        onSelected={vi.fn()}
        onDelete={onDelete}
      />
    );

    const imageElement = screen.getByTestId(`image-${firstImage.id}`);
    fireEvent.mouseEnter(imageElement);

    const deleteIcon = screen.getByTestId(
      `icon-remove-circle-${firstImage.id}`
    );

    fireEvent.click(deleteIcon);

    await vi.waitFor(() => {
      expect(onDelete).toHaveBeenCalledWith(firstImage);
    });
  });

  test("updates selection on prop change", () => {
    const { rerender } = render(
      <ImageSelector
        images={mockImages}
        value={mockImages[0]!}
        onSelected={vi.fn()}
        onDelete={vi.fn()}
      />
    );
    rerender(
      <ImageSelector
        images={mockImages}
        value={mockImages[1]!}
        onSelected={vi.fn()}
        onDelete={vi.fn()}
      />
    );
  });
});
