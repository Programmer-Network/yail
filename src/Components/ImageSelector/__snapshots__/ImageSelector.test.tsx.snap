// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ImageSelector component > renders correctly - snapshot test 1`] = `
<DocumentFragment>
  <div
    class="grid grid-cols-4 gap-4"
  >
    <div
      class="relative aspect-square h-full w-full min-w-[150px] cursor-pointer bg-cover bg-center bg-no-repeat opacity-70 opacity-100!"
      data-testid="image-1"
      style="background-image: url("http://example.com/image1.jpg");"
    />
    <div
      class="relative aspect-square h-full w-full min-w-[150px] cursor-pointer bg-cover bg-center bg-no-repeat opacity-70"
      data-testid="image-2"
      style="background-image: url("http://example.com/image2.jpg");"
    />
  </div>
</DocumentFragment>
`;
